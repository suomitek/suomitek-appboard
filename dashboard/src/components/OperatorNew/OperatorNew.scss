.disabled {
  color: #5f6369;
  cursor: not-allowed;
}
