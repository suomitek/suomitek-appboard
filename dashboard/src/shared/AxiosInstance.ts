import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  InternalServerError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response) {
        if (err.response.data.message) {
          message = err.response.data.message;
        }
        if (typeof err.response.data === "string") {
          message = err.response.data;
        }
      }

      const dispatchErrorAndLogout = (m: string) => {
        // Global action dispatch to log the user out
        dispatch(actions.auth.authenticationError(m));
        // Expire the session if we are using OIDC tokens and
        // logout either way.
        dispatch(actions.auth.expireSession());
      };
      const response = err.response as AxiosResponse;
      switch (response && response.status) {
        case 401:
          dispatchErrorAndLogout(message);
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          // A 403 directly from the auth proxy requires reauthentication.
          if (Auth.usingOIDCToken() && Auth.is403FromAuthProxy(response)) {
            dispatchErrorAndLogout(message);
          }
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        case 500:
          return Promise.reject(new InternalServerError(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
export const axiosWithAuth = Axios.create();
