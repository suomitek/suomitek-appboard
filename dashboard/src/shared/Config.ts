import axios from "axios";

interface ICluster {
  name: string;
  apiServiceURL: string;
  certificateAuthorityData?: string;
}

export interface IFeatureFlags {
  operators: boolean;
  ui: string;
}

// IConfig is the configuration for Kubeapps
export interface IConfig {
  namespace: string;
  appVersion: string;
  authProxyEnabled: boolean;
  oauthLoginURI: string;
  oauthLogoutURI: string;
  error?: Error;
  featureFlags: IFeatureFlags;
  clusters: ICluster[];
}

export default class Config {
  public static async getConfig() {
    const url = Config.APIEndpoint;
    const { data } = await axios.get<IConfig>(url);

    // Development environment config overrides
    // TODO(miguel) Rename env variable to KUBEAPPS_NAMESPACE once/if we eject create-react-app
    // Currently we are using REACT_APP_* because it's the only way to inject env variables in a sealed setup.
    // Please note that this env variable gets mapped in the run command in the package.json file
    if (process.env.NODE_ENV !== "production") {
      data.appVersion = "DEVEL";
      if (process.env.REACT_APP_KUBEAPPS_NS) {
        data.namespace = process.env.REACT_APP_KUBEAPPS_NS;
      }
    }

    return data;
  }

  private static APIEndpoint: string = "config.json";
}
