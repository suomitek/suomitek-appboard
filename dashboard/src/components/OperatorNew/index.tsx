import OperatorNew from "./OperatorNew";

export default OperatorNew;
