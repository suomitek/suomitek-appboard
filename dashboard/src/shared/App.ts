import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const KUBEOPS_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static async create(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.list(cluster, namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    cluster: string,
    namespace: string,
    releaseName: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(
    cluster: string,
    namespace: string,
    releaseName: string,
    revision: number,
  ) {
    const endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(
    cluster: string,
    namespace: string,
    releaseName: string,
    purge: boolean,
  ) {
    let endpoint = url.kubeops.releases.get(cluster, namespace, releaseName);
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  public static async listApps(cluster: string, namespace?: string, allStatuses?: boolean) {
    let endpoint = namespace
      ? url.kubeops.releases.list(cluster, namespace)
      : url.kubeops.releases.listAll(cluster);
    if (allStatuses) {
      endpoint += "?statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(endpoint);
    return data.data;
  }

  public static async getRelease(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      url.kubeops.releases.get(cluster, namespace, name),
    );
    return data.data;
  }
}
