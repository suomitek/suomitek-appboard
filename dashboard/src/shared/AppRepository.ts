import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(namespace, name));
    return data;
  }

  public static async resync(name: string, namespace: string) {
    const repo = await AppRepository.get(name, namespace);
    repo.spec.resyncRequests = repo.spec.resyncRequests || 0;
    repo.spec.resyncRequests++;
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(namespace, name), repo);
    return data;
  }

  public static async update(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(namespace, name),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(name, namespace),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    name: string,
    namespace: string,
    repoURL: string,
    authHeader: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(namespace),
      {
        appRepository: { name, repoURL, authHeader, customCA, syncJobPodTemplate, registrySecrets },
      },
    );
    return data;
  }

  public static async validate(repoURL: string, authHeader: string, customCA: string) {
    const { data } = await axiosWithAuth.post<any>(url.backend.apprepositories.validate(), {
      appRepository: { repoURL, authHeader, customCA },
    });
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getSelfLink(namespace: string, name?: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories${
      name ? `/${name}` : ""
    }`;
  }
}
