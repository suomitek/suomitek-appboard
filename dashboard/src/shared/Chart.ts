import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(namespace: string, repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(
      URL.api.charts.list(namespace, repo),
    );
    return data.data;
  }

  public static async fetchChartVersions(namespace: string, id: string): Promise<IChartVersion[]> {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(namespace, id),
    );
    return data.data;
  }

  public static async getChartVersion(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(namespace, id, version),
    );
    return data.data;
  }

  public static async getReadme(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getReadme(namespace, id, version),
    );
    return data;
  }

  public static async getValues(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(
      URL.api.charts.getValues(namespace, id, version),
    );
    return data;
  }

  public static async getSchema(namespace: string, id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(
      URL.api.charts.getSchema(namespace, id, version),
    );
    return data;
  }

  public static async listWithFilters(
    namespace: string,
    name: string,
    version: string,
    appVersion: string,
  ) {
    const url = `${
      Chart.APIEndpoint
    }/ns/${namespace}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/assetsvc/v1";
}
