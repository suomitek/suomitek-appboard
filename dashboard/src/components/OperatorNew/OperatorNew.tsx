import * as React from "react";

import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import { RouterAction } from "connected-react-router";
import { IOperatorsStateError } from "../../reducers/operators";
import { Operators } from "../../shared/Operators";
import { IPackageManifest, IPackageManifestChannel } from "../../shared/types";
import { api, app } from "../../shared/url";
import { ErrorSelector } from "../ErrorAlert";
import UnexpectedErrorPage from "../ErrorAlert/UnexpectedErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import OperatorHeader from "../OperatorView/OperatorHeader";

import "./OperatorNew.css";

interface IOperatorNewProps {
  operatorName: string;
  operator?: IPackageManifest;
  getOperator: (namespace: string, name: string) => Promise<void>;
  isFetching: boolean;
  cluster: string;
  namespace: string;
  errors: IOperatorsStateError;
  createOperator: (
    namespace: string,
    name: string,
    channel: string,
    installPlanApproval: string,
    csv: string,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
}

interface IOperatorNewState {
  updateChannel?: IPackageManifestChannel;
  updateChannelGlobal: boolean;
  // Instalation mode: true for global, false for namespaced
  installationModeGlobal: boolean;
  // Approval strategy: true for automatic, false for manual
  approvalStrategyAutomatic: boolean;
}

class OperatorNew extends React.Component<IOperatorNewProps, IOperatorNewState> {
  public state: IOperatorNewState = {
    updateChannelGlobal: false,
    installationModeGlobal: false,
    approvalStrategyAutomatic: true,
  };

  public componentDidMount() {
    const { operatorName, namespace, getOperator } = this.props;
    getOperator(namespace, operatorName);
  }

  public componentDidUpdate(prevProps: IOperatorNewProps) {
    if (prevProps.operator !== this.props.operator && this.props.operator) {
      const defaultChannel = Operators.getDefaultChannel(this.props.operator);
      const global = Operators.global(defaultChannel);
      this.setState({
        updateChannel: defaultChannel,
        updateChannelGlobal: global,
        installationModeGlobal: global,
      });
    }
    if (prevProps.namespace !== this.props.namespace) {
      this.props.getOperator(this.props.namespace, this.props.operatorName);
    }
  }

  public render() {
    const { cluster, isFetching, namespace, operatorName, operator, errors, push } = this.props;
    if (cluster !== "default") {
      return <OperatorNotSupported namespace={namespace} />;
    }
    const {
      updateChannel,
      updateChannelGlobal,
      installationModeGlobal,
      approvalStrategyAutomatic,
    } = this.state;
    const error = errors.fetch || errors.create;
    if (error) {
      return <ErrorSelector error={error} resource={`Operator ${operatorName}`} />;
    }
    if (isFetching || !operator) {
      return <LoadingWrapper />;
    }
    if (!updateChannel) {
      return (
        <UnexpectedErrorPage
          text={`Operator ${operatorName} doesn't define a valid channel. This is needed to extract required info.`}
        />
      );
    }
    const { currentCSVDesc } = updateChannel;
    // It's not possible to install a namespaced operator in the "operators" ns
    const disableInstall = namespace === "operators" && !updateChannelGlobal;
    return (
      <section className="ChartView padding-b-big">
        <OperatorHeader
          id={operator.metadata.name}
          description={currentCSVDesc.displayName}
          icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
          version={currentCSVDesc.version}
          cluster={cluster}
          namespace={namespace}
          provider={operator.status.provider.name}
          namespaced={!updateChannelGlobal}
          push={push}
          hideButton={true}
        />
        <main>
          <div className="container container-fluid">
            <div className="row">
              <div className="col-9">
                <form className="container" onSubmit={this.handleDeploy}>
                  <div className="margin-b-normal">
                    <div>
                      <strong>Update Channel</strong>
                    </div>
                    <div className="margin-b-small">
                      The channel to track and receive updates from.
                    </div>
                    {operator.status.channels.map(channel => {
                      return (
                        <label
                          className="margin-l-big"
                          htmlFor={`operator-channel-${channel.name}`}
                          key={channel.name}
                        >
                          <input
                            type="radio"
                            id={`operator-channel-${channel.name}`}
                            name="channel"
                            checked={updateChannel.name === channel.name}
                            onClick={this.selectChannel(channel.name)}
                          />
                          {channel.name}
                          <br />
                        </label>
                      );
                    })}
                  </div>
                  <div className="margin-b-normal">
                    <div className="margin-b-small">
                      <strong>Installation Mode</strong>
                    </div>
                    <label className="margin-l-big" htmlFor="operator-installation-mode-global">
                      <input
                        type="radio"
                        id="operator-installation-mode-global"
                        name="installation-mode"
                        disabled={!updateChannelGlobal}
                        checked={installationModeGlobal}
                        onClick={this.setInstallationMode(true)}
                      />
                      <span className={!updateChannelGlobal ? "disabled" : ""}>
                        All namespaces on the cluster (default)
                      </span>
                      <br />
                      {!updateChannelGlobal && (
                        <div className="disabled margin-l-enormous">
                          <i>This mode is not supported by this Operator and channel.</i>
                        </div>
                      )}
                    </label>
                    <label className="margin-l-big" htmlFor="operator-installation-mode-namespaced">
                      <input
                        type="radio"
                        id="operator-installation-mode-namespaced"
                        name="installation-mode"
                        checked={!installationModeGlobal}
                        onClick={this.setInstallationMode(false)}
                      />
                      The current namespace: {namespace}
                      <br />
                    </label>
                  </div>
                  <div className="margin-b-normal">
                    <div>
                      <strong>Approval Strategy</strong>
                    </div>
                    <div className="margin-b-small">
                      The strategy to determine either manual or automatic updates.
                    </div>
                    <label className="margin-l-big" htmlFor="operator-update-automatic">
                      <input
                        type="radio"
                        id="operator-update-automatic"
                        name="operator-update"
                        checked={approvalStrategyAutomatic}
                        onClick={this.setApprovalStrategy(true)}
                      />
                      Automatic
                      <br />
                    </label>
                    <label className="margin-l-big" htmlFor="operator-update-manual">
                      <input
                        type="radio"
                        id="operator-update-manual"
                        name="operator-update"
                        checked={!approvalStrategyAutomatic}
                        onClick={this.setApprovalStrategy(false)}
                      />
                      Manual
                      <br />
                    </label>
                  </div>
                  {disableInstall && (
                    <UnexpectedErrorPage
                      title={
                        'It\'s not possible to install a namespaced operator in the "operators" namespace'
                      }
                    />
                  )}
                  <button className="button button-primary" type="submit" disabled={disableInstall}>
                    Submit
                  </button>
                </form>
              </div>
            </div>
          </div>
        </main>
      </section>
    );
  }

  private selectChannel = (channel: string) => {
    const newChannel = this.props.operator?.status.channels.find(ch => ch.name === channel);
    const global = Operators.global(newChannel);
    return () => {
      this.setState({
        updateChannel: newChannel,
        updateChannelGlobal: global,
        installationModeGlobal: global,
      });
    };
  };

  private setInstallationMode = (global: boolean) => {
    return () => {
      this.setState({ installationModeGlobal: global });
    };
  };

  private setApprovalStrategy = (automatic: boolean) => {
    return () => {
      this.setState({ approvalStrategyAutomatic: !this.state.approvalStrategyAutomatic });
    };
  };

  private handleDeploy = async () => {
    const { cluster, namespace, operator, createOperator, push } = this.props;
    const { updateChannel, installationModeGlobal, approvalStrategyAutomatic } = this.state;
    const targetNS = installationModeGlobal ? "operators" : namespace;
    const approvalStrategy = approvalStrategyAutomatic ? "Automatic" : "Manual";
    const deployed = await createOperator(
      targetNS,
      operator!.metadata.name,
      updateChannel!.name,
      approvalStrategy,
      updateChannel!.currentCSV,
    );
    if (deployed) {
      // Success, redirect to operator page
      push(app.operators.list(cluster, namespace));
    }
  };
}

export default OperatorNew;
