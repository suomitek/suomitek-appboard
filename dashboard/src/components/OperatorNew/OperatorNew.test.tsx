import OperatorNotSupported from "components/OperatorList/OperatorsNotSupported";
import { shallow } from "enzyme";
import * as React from "react";
import { NotFoundError } from "../../shared/types";
import UnexpectedErrorPage from "../ErrorAlert/UnexpectedErrorAlert";
import OperatorNew from "./OperatorNew";

const defaultProps = {
  operatorName: "foo",
  getOperator: jest.fn(),
  isFetching: false,
  cluster: "default",
  namespace: "kubeapps",
  push: jest.fn(),
  createOperator: jest.fn(),
  errors: {},
};

const defaultOperator = {
  metadata: {
    name: "foo",
    namespace: "kubeapps",
  },
  status: {
    provider: {
      name: "Kubeapps",
    },
    defaultChannel: "beta",
    channels: [
      {
        name: "beta",
        currentCSV: "foo.1.0.0",
        currentCSVDesc: {
          displayName: "Foo",
          version: "1.0.0",
          description: "this is a testing operator",
          annotations: {
            capabilities: "Basic Install",
            repository: "github.com/kubeapps/kubeapps",
            containerImage: "kubeapps/kubeapps",
            createdAt: "one day",
          },
          installModes: [],
        },
      },
    ],
  },
} as any;

it("displays an alert if rendered for an additional cluster", () => {
  const props = { ...defaultProps, cluster: "other-cluster" };
  const wrapper = shallow(<OperatorNew {...props} />);
  expect(wrapper.find(OperatorNotSupported)).toExist();
});

it("calls getOperator when mounting the component", () => {
  const getOperator = jest.fn();
  shallow(<OperatorNew {...defaultProps} getOperator={getOperator} />);
  expect(getOperator).toHaveBeenCalledWith(defaultProps.namespace, defaultProps.operatorName);
});

it("calls getOperator when changing the namespace the component", () => {
  const getOperator = jest.fn();
  const wrapper = shallow(<OperatorNew {...defaultProps} getOperator={getOperator} />);
  expect(getOperator).toHaveBeenCalledWith(defaultProps.namespace, defaultProps.operatorName);
  wrapper.setProps({ namespace: "foo" });
  expect(getOperator).toHaveBeenCalledWith("foo", defaultProps.operatorName);
});

it("parses the default channel when receiving the operator", () => {
  const wrapper = shallow(<OperatorNew {...defaultProps} />);
  wrapper.setProps({ operator: defaultOperator });
  expect(wrapper.state()).toMatchObject({
    updateChannel: defaultOperator.status.channels[0],
    updateChannelGlobal: false,
    installationModeGlobal: false,
  });
});

it("renders an error if present", () => {
  const wrapper = shallow(
    <OperatorNew {...defaultProps} errors={{ fetch: new NotFoundError() }} />,
  );
  expect(wrapper.html()).toContain("Operator foo not found");
});

it("shows an error if the operator doesn't have any channel defined", () => {
  const operator = {
    status: {
      channels: [],
    },
  };
  const wrapper = shallow(<OperatorNew {...defaultProps} operator={operator as any} />);
  expect(
    wrapper
      .find(UnexpectedErrorPage)
      .dive()
      .text(),
  ).toContain(
    "Operator foo doesn't define a valid channel. This is needed to extract required info",
  );
});

it("renders a full OperatorNew", () => {
  const wrapper = shallow(<OperatorNew {...defaultProps} />);
  wrapper.setProps({ operator: defaultOperator });
  expect(wrapper).toMatchSnapshot();
});

it("disables the submit button if the operators ns is selected", () => {
  const wrapper = shallow(<OperatorNew {...defaultProps} namespace="operators" />);
  wrapper.setProps({ operator: defaultOperator });
  expect(wrapper.find("button").prop("disabled")).toBe(true);
  expect(wrapper.find(UnexpectedErrorPage)).toExist();
});

it("deploys an operator", async () => {
  const createOperator = jest.fn().mockReturnValue(true);
  const push = jest.fn();
  const wrapper = shallow(
    <OperatorNew
      {...defaultProps}
      namespace="operators"
      createOperator={createOperator}
      push={push}
    />,
  );
  wrapper.setProps({ operator: defaultOperator });
  const onSubmit = wrapper.find("form").prop("onSubmit") as () => Promise<void>;
  await onSubmit();

  expect(createOperator).toHaveBeenCalledWith("operators", "foo", "beta", "Automatic", "foo.1.0.0");
  expect(push).toHaveBeenCalledWith("/c/default/ns/operators/operators");
});
