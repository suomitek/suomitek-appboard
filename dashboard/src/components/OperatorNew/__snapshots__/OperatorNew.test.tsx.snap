// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a full OperatorNew 1`] = `
<section
  className="ChartView padding-b-big"
>
  <OperatorHeader
    cluster="default"
    description="Foo"
    hideButton={true}
    icon="api/v1/namespaces/kubeapps/operator/foo/logo"
    id="foo"
    namespace="kubeapps"
    namespaced={true}
    provider="Kubeapps"
    push={[MockFunction]}
    version="1.0.0"
  />
  <main>
    <div
      className="container container-fluid"
    >
      <div
        className="row"
      >
        <div
          className="col-9"
        >
          <form
            className="container"
            onSubmit={[Function]}
          >
            <div
              className="margin-b-normal"
            >
              <div>
                <strong>
                  Update Channel
                </strong>
              </div>
              <div
                className="margin-b-small"
              >
                The channel to track and receive updates from.
              </div>
              <label
                className="margin-l-big"
                htmlFor="operator-channel-beta"
                key="beta"
              >
                <input
                  checked={true}
                  id="operator-channel-beta"
                  name="channel"
                  onClick={[Function]}
                  type="radio"
                />
                beta
                <br />
              </label>
            </div>
            <div
              className="margin-b-normal"
            >
              <div
                className="margin-b-small"
              >
                <strong>
                  Installation Mode
                </strong>
              </div>
              <label
                className="margin-l-big"
                htmlFor="operator-installation-mode-global"
              >
                <input
                  checked={false}
                  disabled={true}
                  id="operator-installation-mode-global"
                  name="installation-mode"
                  onClick={[Function]}
                  type="radio"
                />
                <span
                  className="disabled"
                >
                  All namespaces on the cluster (default)
                </span>
                <br />
                <div
                  className="disabled margin-l-enormous"
                >
                  <i>
                    This mode is not supported by this Operator and channel.
                  </i>
                </div>
              </label>
              <label
                className="margin-l-big"
                htmlFor="operator-installation-mode-namespaced"
              >
                <input
                  checked={true}
                  id="operator-installation-mode-namespaced"
                  name="installation-mode"
                  onClick={[Function]}
                  type="radio"
                />
                The current namespace: 
                kubeapps
                <br />
              </label>
            </div>
            <div
              className="margin-b-normal"
            >
              <div>
                <strong>
                  Approval Strategy
                </strong>
              </div>
              <div
                className="margin-b-small"
              >
                The strategy to determine either manual or automatic updates.
              </div>
              <label
                className="margin-l-big"
                htmlFor="operator-update-automatic"
              >
                <input
                  checked={true}
                  id="operator-update-automatic"
                  name="operator-update"
                  onClick={[Function]}
                  type="radio"
                />
                Automatic
                <br />
              </label>
              <label
                className="margin-l-big"
                htmlFor="operator-update-manual"
              >
                <input
                  checked={false}
                  id="operator-update-manual"
                  name="operator-update"
                  onClick={[Function]}
                  type="radio"
                />
                Manual
                <br />
              </label>
            </div>
            <button
              className="button button-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </main>
</section>
`;
